capi-adoption/README.md

```diff
@@ -0,0 +1,114 @@
+
+## Present state of Giant Swarm platform
+
+The present state of Giant Swarm platform consists of a single management cluster on each installation. All workload clusters are managed from the same management cluster on a given installation.
+
+Different release versions of workload clusters are managed by running sets of releases on every management cluster. Each release consists of a set of operators which are reconciling CRs that are labeled to use that given release. Most of the time this worked pretty well for us, given the cost savings that it brought. It also helped us to take control of the lifecycle of the different clusters living on those management clusters, allowing us to upgrade every cluster independently by just changing the label in the CRs.
+
+We were able to utilize a small number of nodes, run our own configuration of kubernetes and still manage a big number of workload clusters run on different versions.
+
+![Multi-version state](multiple_versions.png)
+
+## Problems with the Giant Swarm approach
+
+### Versioning of operators with labels
+
+Our current approach to running multiple versions of the same operator on the same cluster is based on CR labels. This has worked very well until we started deploying admission controllers and later deploying upstream controllers for CAPx.
+
+The nature of upstream CAPI is such that a single workload cluster consists of several different CRs. There can be easily 20 of them and they all must share interrelated references between each other. Some of these are automatically created as part of the reconciliation and this creates big challenges for catching the event at the right point and applying the right Giant Swarm specific labels so that versioned operators can find them. For example, during our upgrade process it’s hard to validate the contents of the labels (is it the wrong value or are we in the middle of an upgrade?) or even determine the order in which the labels must be updated so that the right operators pick up the CRs.
+
+Another problem with label based versioning is that because all code in the upstream ecosystem has been built on the idea of having a single version of an operator on a given cluster, there are several optimizations that leverage this fact. For example, kubebuilder (a widely used framework) has internal re-queuing logic that will repeat reconciliation of a given CR at certain states and this bypasses completely any label filtering. Whether it is built as a label selector for watch operation or post-processing predicate filter as is the current situation with our forked controllers. This has a consequence that change of version label on any given CR does not guarantee that the change is immediately reflected on which operators are reconciling it. In fact, there would often be multiple of them which is very risky.
```

Comment:
I would like to know the technical details of this. Independent of this proposal.

Reply:
As far as I can read code, the predicate filtering is only done on event entry (i.e. when receiving event from informer). 

When re-queuing event, it bypasses this filtering:
- Main event queue processing loop: https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/internal/controller/controller.go#L212-L215
- That then takes an event from queue and calls reconciliation on it: https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/internal/controller/controller.go#L233-L255

It would require extremely careful programming of operator to have 100% confidence that no execution path causes re-queuing of event, while still taking advantage of that optimization feature in general. ...and despite of this, I'd say there would still be a bug hidden somewhere.

Because of this it is impossible to have a clear cut transfer of reconciliation on a given CR with the label change.